src/main/java/org/apache/sling/jcr/resource/internal/JcrValueMap.java

```diff
@@ -392,6 +429,16 @@ void readFully() {
                     final Property prop = pi.nextProperty();
                     this.cacheProperty(prop);
                 }
+                // make sure primary type is in cache
+                if (!this.cache.containsKey(JcrConstants.JCR_PRIMARYTYPE)) {
+                    readPrimaryType();
+                }
+                // make sure sling:resourceType is in cache
+                if (!this.cache.containsKey(ResourceResolver.PROPERTY_RESOURCE_TYPE)) {
+                    // special handling for the resource type property which according to the API must always be exposed via property sling:resourceType
+                    // use value of jcr:primaryType if sling:resourceType is not set
+                    this.cacheProperty(ResourceResolver.PROPERTY_RESOURCE_TYPE, valueCache.get(JcrConstants.JCR_PRIMARYTYPE), node);
```

Comment:
@kwin , similar as above. isn't this method here intended to make sure primary type and resource type are always cached? in which case the primary type property would only act as fallback?